libimage/pull.go

```diff
@@ -379,6 +381,42 @@ func (r *Runtime) copyFromRegistry(ctx context.Context, ref types.ImageReference
 		return nil, err
 	}
 
+	// Avoid running out of time when running inside a systemd unit by
+	// regularly increasing the timeout.
+	if socketPath, _ := os.LookupEnv("NOTIFY_SOCKET"); socketPath != "" {
+		socketAddr := &net.UnixAddr{
+			Name: socketPath,
+			Net:  "unixgram",
+		}
+		conn, err := net.DialUnix(socketAddr.Net, nil, socketAddr)
+		if err != nil {
+			return nil, err
+		}
+		defer conn.Close()
+
+		increaseTimeout := func() {
+			// See upstream systemd change adding support for EXTEND_TIMEOUT_USEC:
+			// https://github.com/systemd/systemd/commit/a327431bd168b2f327f3cd422379e213c643f2a5
+			if _, err := conn.Write([]byte("EXTEND_TIMEOUT_USEC=5000000")); err != nil { // Note it's microseconds.
+				logrus.Errorf("Increasing EXTEND_TIMEOUT_USEC failed: %v", err)
+			}
+		}
+
+		increaseTimeout()
+		socketCtx, cancel := context.WithCancel(ctx)
+		defer cancel()
+		go func() {
+			for {
+				select {
+				case <-socketCtx.Done():
+					return
+				case <-time.After(5 * time.Second):
+					increaseTimeout()
```

Comment:
Should we check if the program is making progress, before increasingTimeout.  If the network connection is hung and will never complete, don't we have the risk of the service never starting?

Reply:
I think can rely on the go standard lib for that.  We'll timeout in `net` in case there's truly no progress.